### ISSUE TITLE
Make ocw-www test course more realistic

### ISSUE THREAD

**Issue by ChristopherChudzicki** [mitodl/ocw-hugo-themes#980]
Right now it is of limited usefulness because it has e.g., no featured courses or stories.

OCW WWW test course should have:
- feaured course cards
- instructors
- build without pointing env vars (e.g., STATIC_API_BASE_URL) to RC

**Comment by ChristopherChudzicki** [comment]
@gumaerc I hit a bit of a snag with this. I was trying to NOT use the rc environment variables in the playwright tests so that `ocw-ci-test-course` could be one of the featured courses in `ocw-ci-test-www`. But it seems to me that:

- building ocw-www requires all featured courses to already be built
- building courses requires ocw-www to already be built (e.g., the course build pulsl instructor data from ocw-www)

So... it seems like there's some circular dependency in the build. I think maybe we would see it in production if you (1) added new instructors, (2) added a new course with the new instructor, (3) made the new course a featured course on www, and (4) then tried to publish both.

Does the above sound accurate?

**Comment by gumaerc** [comment]
@ChristopherChudzicki This situation would be prevented by `ocw-studio` because it does not allow you to assign an instructor to a course if that instructor data hasn't been published. Basically, if there hasn't been a live publish since the instructor was added, then it won't show up in the list

The same goes with featured courses, the won't show up in the relation widget unless they have been built and published live. This does present a conundrum for this particular situation though where we are starting from nothing and want to just run two builds. If this were done in `ocw-studio`, you would build the `ocw-www` site with the instructor but without the featured course and publish it, then publish the featured course with the instructor linked, then add the featured course in `ocw-www` and republish.

**Comment by ChristopherChudzicki** [comment]
> This situation would be prevented by ocw-studio because it does not allow you to assign an instructor to a course if that instructor data hasn't been published. Basically, if there hasn't been a live publish since the instructor was added, then it won't show up in the list

Ah, that makes sense!

I will ponder this conundrum a little more...

**Comment by ChristopherChudzicki** [comment]
<img src="https://i.imgflip.com/715415.jpg" />

**Comment by ChristopherChudzicki** [comment]
@gumaerc Isn't this potentially problematic for our "sync prod db --> rc db, republish all of rc" plan? I guess rc was close enough to prod last time you did the sync that we didn't run into this issue?

**Comment by gumaerc** [comment]
@ChristopherChudzicki There's still work to be done on the DB restore to make it fully automatic, but when I did that for the first time I ran the following in `ocw-studio` RC:

```
./manage.py backpopulate_pipelines
./manage.py reset_sync_states (this will also push all sites to the `main` branch of their github repos, creating if necessary)
./manage.py mass_publish live
```

The thing is, I don't think `ocw-www` will fail to publish if the featured courses aren't published or at least it shouldn't. What's supposed to happen is that the featured courses just don't show up, and then you can republish `ocw-www` when they are available. In the same vein I'm not sure courses should fail to publish if instructors aren't available. We added some error handling when we switched from using Hugo's old `getJSON` function to `resources.GetRemote` but there could be an issue with the error handling there: https://github.com/mitodl/ocw-hugo-themes/pull/734

**Comment by gumaerc** [comment]
Personally, I'm reminded of this image when I think about the situation with external data dependencies in OCW:

![image](https://user-images.githubusercontent.com/12089658/202539378-b06b7646-5c08-4d2b-9ad7-ec4c181619b2.png)

In all seriousness though, this needs to be handled better. We fetch the following data when building the `ocw-www` homepage:
 - Featured Courses (defined in `ocw-www` as a list of UUID's for which data is fetched at build time through the static JSON API)
 - New Courses (fetched through the `ocw-studio` API at the `/api/websites` endpoint filtered on `type` with a value of whatever is set on `OCW_COURSE_STARTER_SLUG`)
 - OCW News (fetched through the `ocw-studio` API at the `/api/news` endpoint
 
In theory we should throw a hard error if fetching any of this information fails, because the home page would be incomplete and should not be deployed if that's so. From that standpoint, that's something we really should be testing here.

Since all of this is basically just JSON data, one idea would be maybe including some JSON test fixtures to make available to Playwright at the time of testing to simulate a "I was able to get all of the data" situation, then also test without them and assert the failures?

**Comment by ChristopherChudzicki** [comment]
> Since all of this is basically just JSON data, one idea would be maybe including some JSON test fixtures to make available to Playwright at the time of testing to simulate a "I was able to get all of the data" situation, then also test without them and assert the failures?

Of course! I will try that.

**Comment by ChristopherChudzicki** [comment]
>... then also test without them and assert the failures?

_Partly as a reminder to myself:_ We should definitely test that sort of failed build situation, but it won't be a playwright test, since we'll never get to the browser. Probably just `jest`.

**Comment by ChristopherChudzicki** [comment]
@gumaerc Just wanted to let you know that we merged this with the mock approach. Now `yarn test:e2e`:

1. Starts a mock static API server for requests like `courses/8.06-quantum-stuff/data.json`
2. runs hugo build, calling out to the static server rather than the build output (since it doesn't exist yet)
3. runs e2e test, including a test that checks the static server mocks are what the build produces.

I did **not** do anything yet regarding build errors / error handling you mentioned in https://github.com/mitodl/ocw-hugo-themes/issues/980#issuecomment-1319126634. I think (1) that's best done in a separate PR, and (2) we should a slightly larger discussion about how errors are handled (e.g., what is fatal and what is not[^1] in OCW, but I need to finish some subscript/superscript stuff first.

[^1]: Though I agree with your inclination homepage errors should be fatal
